pkg/minishift/openshift/service.go

```diff
@@ -192,46 +205,65 @@ func calculateWeight(a float64, b float64) string {
 	return ""
 }
 
+func hasNodePort(serviceNodePort map[string]string, name string) bool {
+	_, ok := serviceNodePort[name]
+	return ok
+}
+
+func createServiceSpec(service ServiceWeight, routeURL, namespace, nodePort string) *ServiceSpec {
+	serviceSpec := &ServiceSpec{Name: service.Name,
+		Namespace: namespace,
+		Weight:    []string{service.Weight},
+		NodePort:  nodePort,
+	}
+
+	serviceSpec.URL = append(serviceSpec.URL, genUrl(service, routeURL))
+	return serviceSpec
+}
+
+func genUrl(service ServiceWeight, url string) string {
+	if service.Tls != "" {
+		return fmt.Sprintf(https, url)
+	}
+	return fmt.Sprintf(http, url)
+}
+
+func NewServiceSpecMap() *ServiceSpecMap {
+	return &ServiceSpecMap{S: make(map[string]*ServiceSpec)}
+}
+
+func (s *ServiceSpecMap) Add(service ServiceWeight, routeURL, namespace, nodePort string) {
+	if s.HasService(service.Name) {
+		s.S[service.Name].URL = append(s.S[service.Name].URL, genUrl(service, routeURL))
+		s.S[service.Name].Weight = append(s.S[service.Name].Weight, service.Weight)
+	} else {
+		s.S[service.Name] = createServiceSpec(service, routeURL, namespace, nodePort)
+	}
+}
+
+func (s *ServiceSpecMap) HasService(name string) bool {
+	_, ok := s.S[name]
+	return ok
+}
+
 func filterAndUpdateServiceSpecs(routeSpec map[string][]ServiceWeight, namespace string, serviceNodePort map[string]string) []ServiceSpec {
 	var serviceSpecs []ServiceSpec
-	serviceSpecMap := make(map[string]*ServiceSpec)
-
+	serviceSpecMap := NewServiceSpecMap()
 	for routeURL, v := range routeSpec {
 		for _, service := range v {
-			// split content on colon to separate NAME, HOST and Weight
-			if _, ok := serviceNodePort[service.Name]; ok {
-				if _, ok := serviceSpecMap[service.Name]; !ok {
-					serviceSpecMap[service.Name] = &ServiceSpec{Namespace: namespace,
-						Name:     service.Name,
-						URL:      []string{fmt.Sprintf("http://%s", routeURL)},
-						NodePort: serviceNodePort[service.Name],
-						Weight:   []string{service.Weight},
-					}
-				} else {
-					serviceSpecMap[service.Name].URL = append(serviceSpecMap[service.Name].URL, fmt.Sprintf("http://%s", routeURL))
-					serviceSpecMap[service.Name].Weight = append(serviceSpecMap[service.Name].Weight, service.Weight)
-				}
+			if hasNodePort(serviceNodePort, service.Name) {
```

Comment:
`serviceNodePort` is a little bit unfortunately named. It almost feels, because of the name, that THIS test will alwats return `true` anyway.

Reply:
the origin of the name is:
```
	serviceNodePort, err := getService(namespace)
```
and I think this can be renamed to `service`. or if that is too generic, `serviceForNamespace`, etc.